server/src/managers/websocket/websocket_manager.rs

```diff
@@ -0,0 +1,366 @@
+use axum::extract::Query;
+use axum::http::{StatusCode, Uri};
+use axum::routing::head;
+use libsignal_core::{ProtocolAddress};
+use serde::de::IntoDeserializer;
+use serde::Deserialize;
+use serde_json::json;
+use sha2::digest::consts::False;
+use sha2::digest::typenum::Integer;
+use std::fmt::{self, format, Debug};
+use std::future::Future;
+use std::net::SocketAddr;
+use tonic::ConnectError;
+use uuid::fmt::Braced;
+
+use std::str::FromStr;
+use std::thread::current;
+use std::{collections::HashMap, sync::Arc};
+use tokio::sync::{Mutex, MutexGuard, RwLock};
+
+use common::signal_protobuf::{
+    envelope, web_socket_message, Envelope, WebSocketMessage, WebSocketRequestMessage,
+    WebSocketResponseMessage,
+};
+use common::web_api::{SignalMessage, SignalMessages};
+use prost::{bytes::Bytes, Message as PMessage};
+
+use rand::rngs::OsRng;
+use rand::Rng;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use url::Url;
+
+use crate::account::Account;
+use crate::database::SignalDatabase;
+use crate::managers::state::SignalServerState;
+use super::connection::{WSStream, WebSocketConnection, ConnectionState, ConnectionMap, ClientConnection};
+use crate::error::{ApiError, SocketManagerError};
+use crate::query::PutV1MessageParams;
+
+use chrono::Utc;
+
+/*const WS_ENDPOINTS: [&str; 24] = [
+    "v4/attachments/form/upload",
+    "v1/config",
+    "v1/certificate/delivery",
+    "v1/devices",
+    "v2/directory/auth",
+    "v1/certificate/auth/group",
+    "v1/archives/auth",
+    "v1/archives/auth/read",
+    "v1/archives/upload/form",
+    "v1/archives/media/upload/form",
+    "v1/devices/link",
+    "v1/messages",
+    "v1/messages/multi_recipient",
+    "v2/accounts/phone_number_discoverability",
+    "v1/profile",
+    "v1/archives",
+    "v1/archives/media",
+    "v1/archives/media/batch",
+    "v1/archives/media/delete",
+    "v1/devices/capabilities",
+    "v1/messages/report",
+    "v1/archives/backupid",
+    "v1/archives/keys",
+    "v1/storage/auth",
+];*/
+
+
+
+#[derive(Debug)]
+pub struct WebSocketManager<T: WSStream + Debug> {
+    sockets: ConnectionMap<T>,
+}
+
+impl<T: WSStream + Debug> Clone for WebSocketManager<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sockets: Arc::clone(&self.sockets),
+        }
+    }
+}
+
+impl <T: WSStream + Debug + Send + 'static> WebSocketManager<T> {
+
+    pub fn new() -> Self {
+        Self {
+            sockets: Arc::new(Mutex::new(HashMap::new()))
+        }
+    }
+
+    pub async fn insert<U: SignalDatabase + Send + 'static>(&mut self, connection: WebSocketConnection<T>, state: SignalServerState<U>){
+        let address = connection.protocol_address();
+        let connection: ClientConnection<T> = Arc::new(Mutex::new(connection));
+        
+        self.sockets.lock().await.insert(address.clone(), connection.clone());
+        let mut mgr = self.clone();
+        
+        let protocol_address = address.clone();
```

Comment:
Maybe just call address protocol_address, instead of assigning twice

Reply:
ok